In [ ]:
import pandas as pd

df = pd.read_csv("hf://datasets/scikit-learn/churn-prediction/dataset.csv")

In [ ]:
'''Customer churn prediction dataset of a fictional telecommunication company made by IBM Sample Datasets. Context Predict behavior to retain customers. You can analyze all relevant customer data and develop focused customer retention programs.

Content Each row represents a customer, each column contains customer’s attributes described on the column metadata.

The data set includes information about:

Customers who left within the last month: the column is called Churn
Services that each customer has signed up for: phone, multiple lines, internet, online security, online backup, device protection, tech support, and streaming TV and movies
Customer account information: how long they’ve been a customer, contract, payment method, paperless billing, monthly charges, and total charges
Demographic info about customers: gender, age range, and if they have partners and dependents'''

In [ ]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [ ]:
missing_value = df.isnull().sum().sum()/(df.shape[0]*df.shape[1])*100

In [ ]:
print(missing_value)

0.0


In [ ]:
df.TotalCharges = pd.to_numeric(df.TotalCharges, errors = 'coerce')

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [ ]:
len(df)

7043

In [ ]:
df.dropna(how = "any",inplace = True)

In [ ]:
len(df)

7032

In [ ]:
df.Churn.value_counts()/len(df) *100

,count
Churn,
No,73.421502
Yes,26.578498


In [ ]:
df.columns

Index(['customerID', 'gender', 'SeniorCitizen', 'Partner', 'Dependents',
       'tenure', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod', 'MonthlyCharges', 'TotalCharges', 'Churn'],
      dtype='object')

In [ ]:
x = df.drop(['customerID','Churn'],axis = 1)

In [ ]:
y = df.Churn.values

In [ ]:
print(len(x),len(y))

7032 7032


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7032 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7032 non-null   object 
 1   gender            7032 non-null   object 
 2   SeniorCitizen     7032 non-null   int64  
 3   Partner           7032 non-null   object 
 4   Dependents        7032 non-null   object 
 5   tenure            7032 non-null   int64  
 6   PhoneService      7032 non-null   object 
 7   MultipleLines     7032 non-null   object 
 8   InternetService   7032 non-null   object 
 9   OnlineSecurity    7032 non-null   object 
 10  OnlineBackup      7032 non-null   object 
 11  DeviceProtection  7032 non-null   object 
 12  TechSupport       7032 non-null   object 
 13  StreamingTV       7032 non-null   object 
 14  StreamingMovies   7032 non-null   object 
 15  Contract          7032 non-null   object 
 16  PaperlessBilling  7032 non-null   object 
 17  

In [ ]:
x = pd.get_dummies(x, columns = ['gender',  'Partner', 'Dependents', 'PhoneService', 'MultipleLines', 'InternetService',
       'OnlineSecurity', 'OnlineBackup', 'DeviceProtection', 'TechSupport',
       'StreamingTV', 'StreamingMovies', 'Contract', 'PaperlessBilling',
       'PaymentMethod'], drop_first= True)

In [ ]:
x.head(5)

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges,gender_Male,Partner_Yes,Dependents_Yes,PhoneService_Yes,MultipleLines_No phone service,MultipleLines_Yes,...,StreamingTV_No internet service,StreamingTV_Yes,StreamingMovies_No internet service,StreamingMovies_Yes,Contract_One year,Contract_Two year,PaperlessBilling_Yes,PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
0,0,1,29.85,29.85,False,True,False,False,True,False,...,False,False,False,False,False,False,True,False,True,False
1,0,34,56.95,1889.50,True,False,False,True,False,False,...,False,False,False,False,True,False,False,False,False,True
2,0,2,53.85,108.15,True,False,False,True,False,False,...,False,False,False,False,False,False,True,False,False,True
3,0,45,42.30,1840.75,True,False,False,False,True,False,...,False,False,False,False,True,False,False,False,False,False
4,0,2,70.70,151.65,False,False,False,True,False,False,...,False,False,False,False,False,False,True,False,True,False


In [ ]:
x = x.astype(int)

In [ ]:
x.head(5)

,SeniorCitizen,tenure,MonthlyCharges,TotalCharges,gender_Male,Partner_Yes,Dependents_Yes,PhoneService_Yes,MultipleLines_No phone service,MultipleLines_Yes,...,StreamingTV_No internet service,StreamingTV_Yes,StreamingMovies_No internet service,StreamingMovies_Yes,Contract_One year,Contract_Two year,PaperlessBilling_Yes,PaymentMethod_Credit card (automatic),PaymentMethod_Electronic check,PaymentMethod_Mailed check
0,0,1,29,29,0,1,0,0,1,0,...,0,0,0,0,0,0,1,0,1,0
1,0,34,56,1889,1,0,0,1,0,0,...,0,0,0,0,1,0,0,0,0,1
2,0,2,53,108,1,0,0,1,0,0,...,0,0,0,0,0,0,1,0,0,1
3,0,45,42,1840,1,0,0,0,1,0,...,0,0,0,0,1,0,0,0,0,0
4,0,2,70,151,0,0,0,1,0,0,...,0,0,0,0,0,0,1,0,1,0


In [ ]:
from sklearn.model_selection import train_test_split

In [ ]:
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.25,stratify=y,random_state=42)

In [ ]:
len(x_train)

5274

In [ ]:
len(x_test)

1758

In [ ]:
from sklearn.preprocessing import StandardScaler

In [ ]:
# Standardize features to zero mean and unit variance.

sc = StandardScaler()
x_train_sc= sc.fit_transform(x_train)
x_test_sc= sc.transform(x_test)

** KNN classifier **

In [ ]:
from sklearn.neighbors import KNeighborsClassifier

In [ ]:
# KNN is used as a non-parametric baseline to capture local patterns
# Distance-based models are sensitive to feature scale, so StandardScaler was applied
knn_model = KNeighborsClassifier(n_neighbors=5)

knn_model.fit(x_train_sc, y_train)

# Predictions will be evaluated using ROC-AUC and recall
y_pred_knn = knn_model.predict(x_test_sc)


In [ ]:
#checck accuracy
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,y_pred_knn)*100)

74.63026166097838


In [ ]:
from sklearn.linear_model import LogisticRegression

In [ ]:
model_lr = LogisticRegression()
model_lr.fit(x_train_sc,y_train)
y_pred_lr = model_lr.predict(x_test_sc)

In [ ]:
print(accuracy_score(y_test,y_pred_lr)*100)

80.60295790671216


** Random forst **

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
model_rm  = RandomForestClassifier(n_estimators=200)
model_rm.fit(x_train_sc,y_train)
y_pred_rm = model_rm.predict(x_test_sc)


In [ ]:
print(accuracy_score(y_test,y_pred_rm))



0.7849829351535836


**xg boost**

In [ ]:
from sklearn.metrics import classification_report
knn_model = classification_report(y_test, y_pred_knn)
model_lr = classification_report(y_test, y_pred_lr)
model_rm = classification_report(y_test, y_pred_rm)


In [ ]:
print(knn_model,model_lr,model_rm)

              precision    recall  f1-score   support

          No       0.83      0.83      0.83      1291
         Yes       0.52      0.51      0.52       467

    accuracy                           0.75      1758
   macro avg       0.67      0.67      0.67      1758
weighted avg       0.74      0.75      0.75      1758
               precision    recall  f1-score   support

          No       0.85      0.89      0.87      1291
         Yes       0.65      0.57      0.61       467

    accuracy                           0.81      1758
   macro avg       0.75      0.73      0.74      1758
weighted avg       0.80      0.81      0.80      1758
               precision    recall  f1-score   support

          No       0.83      0.89      0.86      1291
         Yes       0.62      0.49      0.55       467

    accuracy                           0.78      1758
   macro avg       0.72      0.69      0.70      1758
weighted avg       0.77      0.78      0.78      1758



In [ ]:

y[y=='No'] = 0
y[y=='Yes'] = 1
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25)

# Converting data to integer
y_test = y_test.astype(int)
y_train = y_train.astype(int)

# Feature Scaling
from sklearn.preprocessing import StandardScaler

sc = StandardScaler()
x_train_sc = sc.fit_transform(x_train)
x_test_sc = sc.transform(x_test)

In [ ]:
from xgboost import XGBRFClassifier

model_xg = XGBRFClassifier(n_estimators = 200,max_depth = 4 ,learning_rate = 0.1)
model_xg.fit(x_train_sc,y_train)
y_pred_xg = model_xg.predict(x_test_sc)
print(accuracy_score(y_test,y_pred_xg)*100)

72.52559726962457


In [ ]:
model_xg = classification_report(y_test, y_pred_xg)

/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))
/usr/local/lib/python3.12/dist-packages/sklearn/metrics/_classification.py:1565: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


In [ ]:
print(model_xg)

              precision    recall  f1-score   support

           0       0.73      1.00      0.84      1275
           1       0.00      0.00      0.00       483

    accuracy                           0.73      1758
   macro avg       0.36      0.50      0.42      1758
weighted avg       0.53      0.73      0.61      1758



#kNN → weak recall for churn (0.51)

#Random Forest → recall dropped further (0.49)

#XGBoost (current) → ❌ completely failed (predicts only “No”)
# Logistic Regression is selected as the final model because:
# 1. The dataset is imbalanced, and recall for churned customers is critical
# 2. Logistic Regression achieved the highest recall and F1-score for the 'Yes' (churn) class
# 3. The model is interpretable and suitable for business decision-making
# 4. XGBoost failed due to class imbalance, and Random Forest underperformed on churn recall


In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report,confusion_matrix,roc_auc_score

In [ ]:
final_model = LogisticRegression(
    class_weight='balanced',
    max_iter=1000,
    random_state = 42
)
final_model.fit(x_train_sc,y_train)


LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)

In [ ]:
y_prob = final_model.predict_proba(x_test_sc)[:,1]
print(y_prob)

[0.29636445 0.40202795 0.88563735 ... 0.78968625 0.21584599 0.44578608]


In [ ]:
# threshold tuning (lower than 0.5 to catch churn)
y_pred_final = (y_prob >= 0.35).astype(int)

In [ ]:
print(accuracy_score(y_test,y_pred_final))

0.6916951080773607


In [ ]:
print("Final Logistic Regression Model")
print(classification_report(y_test, y_pred_final))

print("Confusion Matrix:")
print(confusion_matrix(y_test, y_pred_final))

print("ROC-AUC Score:", roc_auc_score(y_test, y_prob))

Final Logistic Regression Model
              precision    recall  f1-score   support

           0       0.95      0.61      0.74      1275
           1       0.47      0.92      0.62       483

    accuracy                           0.69      1758
   macro avg       0.71      0.76      0.68      1758
weighted avg       0.82      0.69      0.71      1758

Confusion Matrix:
[[774 501]
 [ 41 442]]
ROC-AUC Score: 0.8562611131409085


Recall = 0.92 % model correctly identifies 92% of customers who will churn.

F1-score = 0.62

ROC-AUC = 0.856

[[774 501]   - Non-churn

 [ 41 442]]  - Churn

 442 churners correctly identified
41 churners missed (very low — GOOD)
501 false positives (acceptable trade-off)

The dataset was imbalanced, so accuracy was not a reliable metric.
I optimized the model for recall on the churn class because missing a churn customer is more costly than falsely flagging a loyal one.
By applying class weighting and threshold tuning, the final Logistic Regression model achieved a recall of 92% and a ROC-AUC of 0.85, making it effective for proactive customer retention.


The dataset was imbalanced, so accuracy was not a reliable metric.
I optimized the model for recall on the churn class because missing a churn customer is more costly than falsely flagging a loyal one.
By applying class weighting and threshold tuning, the final Logistic Regression model achieved a recall of 92% and a ROC-AUC of 0.85, making it effective for proactive customer retention.

clear  conclusion -
The objective of this project was to identify customers at high risk of churn so that proactive retention strategies could be applied. Using customer demographic, service usage, and account information, a Logistic Regression model was developed and optimized to prioritize the detection of churned customers.

Due to the imbalanced nature of the dataset, accuracy alone was not sufficient for evaluation. The model was therefore optimized for recall of churned customers, as failing to identify a customer who is about to leave has a higher business cost than incorrectly flagging a loyal customer.

The final model achieved:

92% recall for churned customers

ROC-AUC score of 0.85, indicating strong class separation ability

This means the model is able to correctly identify the majority of customers who are likely to churn, enabling the business to intervene early with retention offers, improved service, or personalized engagement.

Although the model produces some false positives, this trade-off is acceptable in a churn prediction scenario, as the cost of offering retention incentives to some loyal customers is significantly lower than losing high-value customers.

Business Impact-

Enables early identification of at-risk customers

Supports data-driven retention strategies

Helps reduce customer attrition and revenue loss

Provides an interpretable model suitable for business decision-making

Recommendations for the Business-

Target customers predicted as high churn risk with retention campaigns

Focus on customers with short tenure and flexible contracts, as they show higher churn tendency

Continuously retrain the model as customer behavior and service offerings evolve

In [ ]:
feature_importance = pd.DataFrame({
    "Feature": x.columns,
    "Coefficient": final_model.coef_[0]
}).sort_values(by="Coefficient", ascending=False)

print(feature_importance)

                                  Feature  Coefficient
10            InternetService_Fiber optic     0.689553
3                            TotalCharges     0.566965
23                    StreamingMovies_Yes     0.214487
26                   PaperlessBilling_Yes     0.179030
21                        StreamingTV_Yes     0.174409
28         PaymentMethod_Electronic check     0.158218
9                       MultipleLines_Yes     0.150559
0                           SeniorCitizen     0.069430
17                   DeviceProtection_Yes     0.039695
8          MultipleLines_No phone service     0.025969
5                             Partner_Yes     0.004679
15                       OnlineBackup_Yes    -0.003466
29             PaymentMethod_Mailed check    -0.013326
4                             gender_Male    -0.019084
7                        PhoneService_Yes    -0.025969
27  PaymentMethod_Credit card (automatic)    -0.046993
19                        TechSupport_Yes    -0.075351
18        

In [ ]:
# final_model.coef_[0] → learned weights for each feature

**For client -Customers most likely to leave**

Fiber optic users show a higher tendency to leave, possibly due to higher costs or stronger market competition.

Customers with higher overall bills are more price-sensitive and more likely to churn.

Users of streaming TV and movie services tend to switch providers more frequently.

Customers using paperless billing or electronic check payments are generally less tied to long-term commitments.

Customers with multiple phone lines or senior citizens may experience higher service complexity, increasing dissatisfaction.

**Customers more likely to stay**

Long-term customers are significantly more loyal; churn risk is highest during the early months.

Customers with lower monthly charges show stronger retention.

Long-term contracts (1-year or 2-year) greatly reduce churn.

Customers subscribed to technical support and security services are more satisfied and loyal.

Customers with dependents and automatic payments tend to remain with the company longer.

The analysis shows that churn is highest among customers with flexible contracts, higher charges, and premium services such as fiber optic and streaming. In contrast, long-term contracts, lower pricing, and value-added services like tech support significantly improve retention. Focusing on early-stage customers and offering contract upgrades and bundled services can meaningfully reduce churn.

how to read feature --

1️⃣ How to read this feature-importance table

Positive coefficient (+) → increases churn risk

Negative coefficient (−) → reduces churn risk

Bigger absolute value → stronger influence

Your model learned real behavioral patterns — this is not random.

2️⃣ Top drivers that INCREASE churn (high-risk customers)
🚨 Strong churn indicators

1️⃣ Fiber optic internet (+0.69)
Customers using fiber optic services are significantly more likely to churn.
👉 Possible reasons: higher cost, service issues, stronger competition.

2️⃣ High total charges (+0.57)
Customers who have paid more overall show higher churn tendency.
👉 Indicates price sensitivity or dissatisfaction with value.

3️⃣ Streaming services (+0.21, +0.17)
Streaming TV and movies users churn more.
👉 Likely using telecom mainly for entertainment and switching providers easily.

4️⃣ Paperless billing & electronic check payments (+0.18, +0.16)
Digitally paying customers show higher churn.
👉 These customers are more flexible and less contract-bound.

5️⃣ Multiple phone lines & senior citizens (moderate impact)
Indicates complexity of services may increase dissatisfaction.

3️⃣ Features that REDUCE churn (loyal customers)
🛡️ Strong retention signals

1️⃣ Long tenure (−1.23) — strongest factor
The longer a customer stays, the less likely they churn.
👉 Early customer period is the most critical.

2️⃣ Lower monthly charges (−0.71)
Affordable pricing strongly improves retention.

3️⃣ Long-term contracts

Two-year contract (−0.62)

One-year contract (−0.27)

👉 Contract commitment dramatically reduces churn.

4️⃣ Online security & tech support (−0.14, −0.07)
Customers with support services are more loyal.

5️⃣ Dependents & automatic credit card payments
Stable households and automated payments reduce churn risk.


